imp_developer.py

```diff
@@ -272,56 +326,414 @@ def __base64_encode(str):
             result = json.loads(err.read().decode('utf-8'))
+        except urllib.error.URLError as err:
+            code = 404
+            result = {"error" : err.reason}
+        except urllib.error.ContentTooShortError as err:
+            code = 404
+            result = {"error" : "Too short conent exception"}
+
         return result, code
 
     @staticmethod
-    def get(key, url, timeout=None):
-        return HTTP.do_request(key, url, "GET", timeout=timeout)
+    def get(key, url, timeout=None, data=None, headers=HttpHeaders.DEFAULT):
+        return HTTP.do_request(key, url, "GET", timeout=timeout, data=data, headers=headers)
 
     @staticmethod
-    def post(key, url, data=None):
-        return HTTP.do_request(key, url, "POST", data)
+    def post(key, url, data=None, headers=HttpHeaders.DEFAULT):
+        return HTTP.do_request(key, url, "POST", data, headers=headers)
 
     @staticmethod
-    def put(key, url, data=None):
-        return HTTP.do_request(key, url, "PUT", data)
+    def put(key, url, data=None, headers=HttpHeaders.DEFAULT):
+        return HTTP.do_request(key, url, "PUT", data, headers=headers)
+
+    @staticmethod
+    def delete(key, url, data=None, headers=HttpHeaders.DEFAULT):
+        return HTTP.do_request(key, url, "DELETE", data, headers=headers)
 
     @staticmethod
     def is_response_code_valid(code):
-        return code in [200, 201, 202]
+        return code in [200, 201, 202, 204]
+
+    @staticmethod
+    def is_invalid_credentials(code, response):
+        if code in [401]:
+            errors = response.get("errors")
+            if errors and len(errors) > 0:
+                # it is possible to get more than one error
+                # but usually it is only one error
+                for error in errors:
+                    if error.get("title") in ["'Invalid Credentials'"] or error.get("code") == "PX100":
+                        return True
+        return False
+
+    @staticmethod
+    def is_wrong_input(code, response):
+        if code in [400, 409]:
+            errors = response.get("errors")
+            if errors and len(errors) > 0:
+                # it is possible to get more than one error
+                # but usually it is only one error
+                for error in errors:
+                    if error.get("detail"):
+                        return error.get("detail"), error.get("title")
+        return None, None
+
+    @staticmethod
+    def is_failure_request(response, code):
+        failure = None
+        if not HTTP.is_response_code_valid(code):
+            if code == 404:
+                return "\n There is no Internet connection.\n Or requested resource not avialble."
+            errors = response.get("errors")
+            if errors and len(errors) > 0:
+                # take only first error to show
+                failure = errors[0]["detail"]
+        # return the failure message
+        return failure
+
+class ImpRequest():
```

Comment:
extra empty line before the class definition is required

Reply:
fixed